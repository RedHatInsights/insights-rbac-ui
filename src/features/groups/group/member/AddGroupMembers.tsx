import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { useParams } from 'react-router-dom';
import { useIntl } from 'react-intl';
import { useFlag } from '@unleash/proxy-client-react';
import { Button } from '@patternfly/react-core/dist/dynamic/components/Button';
import { Modal } from '@patternfly/react-core/dist/dynamic/components/Modal';
import { ModalVariant } from '@patternfly/react-core';
import { Stack } from '@patternfly/react-core';
import { StackItem } from '@patternfly/react-core';
import { TextContent } from '@patternfly/react-core/dist/dynamic/components/Text';
import { addNotification } from '@redhat-cloud-services/frontend-components-notifications/';
import { useChrome } from '@redhat-cloud-services/frontend-components/useChrome';
import { addMembersToGroup, fetchGroups, fetchMembersForGroup } from '../../../../redux/groups/actions';
import { UsersList } from '../../add-group/components/stepUsers/UsersList';
import { ActiveUsers } from '../../../../components/user-management/ActiveUsers';
import useAppNavigate from '../../../../hooks/useAppNavigate';
import messages from '../../../../Messages';
import type { AddGroupMembersProps } from './types';

export const AddGroupMembers: React.FC<AddGroupMembersProps> = ({ cancelRoute }) => {
  const chrome = useChrome();
  const intl = useIntl();
  const navigate = useAppNavigate();
  const { groupId } = useParams<{ groupId: string }>();
  const dispatch = useDispatch();
  const isITLess = useFlag('platform.rbac.itless');

  const [selectedUsers, setSelectedUsers] = useState<any[]>([]);

  const onSubmit = () => {
    const userList = selectedUsers.map((user) => ({ username: user.username || user.label }));
    if (userList.length > 0) {
      dispatch(
        addNotification({
          variant: 'info',
          title: intl.formatMessage(userList.length > 1 ? messages.addingGroupMembersTitle : messages.addingGroupMemberTitle),
          description: intl.formatMessage(userList.length > 1 ? messages.addingGroupMembersDescription : messages.addingGroupMemberDescription),
        }) as any,
      );
      (dispatch(addMembersToGroup(groupId!, userList)) as any).then(() => {
        dispatch(fetchMembersForGroup(groupId!));
        dispatch(fetchGroups({ usesMetaInURL: true, chrome }));
      });
    }
    navigate(cancelRoute);
  };

  const onCancel = () => {
    dispatch(
      addNotification({
        variant: 'warning',
        title: intl.formatMessage(messages.addingGroupMembersCancelled),
        description: 'Adding members to group has been cancelled.',
      }) as any,
    );
    navigate(cancelRoute);
  };

  return (
    <Modal
      title={intl.formatMessage(messages.addMembers)}
      variant={ModalVariant.large}
      isOpen
      onClose={onCancel}
      actions={[
        <Button ouiaId="add-members-confirm" key="confirm" variant="primary" onClick={onSubmit} isDisabled={selectedUsers.length === 0}>
          {intl.formatMessage(messages.addToGroup)}
        </Button>,
        <Button ouiaId="add-members-cancel" key="cancel" variant="link" onClick={onCancel}>
          {intl.formatMessage(messages.cancel)}
        </Button>,
      ]}
    >
      <Stack hasGutter>
        <StackItem>
          <TextContent>
            <ActiveUsers {...(!isITLess && { linkDescription: intl.formatMessage(messages.toManageUsersText) })} />
          </TextContent>
        </StackItem>
        <StackItem isFilled>
          {isITLess ? (
            <UsersList initialSelectedUsers={selectedUsers} onSelect={setSelectedUsers} displayNarrow={true} />
          ) : (
            <UsersList initialSelectedUsers={selectedUsers} onSelect={setSelectedUsers} displayNarrow={true} />
          )}
        </StackItem>
      </Stack>
    </Modal>
  );
};

export default AddGroupMembers;
