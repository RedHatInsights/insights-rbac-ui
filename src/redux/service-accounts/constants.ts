export const NO_DATA = 'no-data';
export const LAST_PAGE = 'last-page';
export const RESULTS = 'results';
export const ERROR = 'error';
