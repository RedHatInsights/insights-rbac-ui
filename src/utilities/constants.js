export const RBAC_API_BASE = `/api/rbac/v1`;
