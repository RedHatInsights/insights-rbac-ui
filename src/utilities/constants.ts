export const RBAC_API_BASE = `/api/rbac/v1` as const;
export const RBAC_API_BASE_2 = `/api/rbac/v2` as const;
export const COST_API_BASE = `/api/cost-management/v1` as const;
export const INVENTORY_API_BASE = `/api/inventory/v1` as const;

export const DEFAULT_MUA_BUNDLE = 'rhel' as const;

export const DEFAULT_ACCESS_GROUP_ID = 'default-access' as const;
