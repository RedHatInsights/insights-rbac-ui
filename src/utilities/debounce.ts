import awesomeDebouncePromise from 'awesome-debounce-promise';
import type { AwesomeDebounceOptions } from 'awesome-debounce-promise';
import lodashDebounce from 'lodash/debounce';
import type { DebounceSettings } from 'lodash';

/**
 * Default debounce delay in milliseconds
 * 250ms is the sweet spot - fast enough to feel responsive, slow enough to reduce API calls
 */
export const DEFAULT_DEBOUNCE_DELAY = 250;

/**
 * Debounces an async function with race condition prevention
 *
 * @param fn - The async function to debounce
 * @param wait - Delay in milliseconds
 * @param options - Must include onlyResolvesLast: true for race-safe debouncing
 * @returns Debounced version of the function with cancel() method
 *
 * @example
 * const safeFetch = debounce(fetchId, 300, { onlyResolvesLast: true });
 */
export function debounce<T extends (...args: any[]) => Promise<any>>(
  fn: T,
  wait: number,
  options: AwesomeDebounceOptions & { onlyResolvesLast: true },
): T & { cancel(): void };

/**
 * Debounces a function (sync or async without race guards)
 *
 * @param fn - The function to debounce (sync or async)
 * @param wait - Delay in milliseconds (default: 250ms)
 * @param options - Standard lodash debounce options
 * @returns Debounced version of the function with cancel() and flush() methods
 *
 * @example
 * // Async functions
 * const debouncedSearch = debounce(
 *   async (query: string) => fetchResults(query)
 * );
 *
 * // Redux dispatch
 * const debouncedFetch = debounce(fetchData);
 *
 * // Cleanup on unmount
 * useEffect(() => () => debouncedFetch.cancel(), []);
 */
export function debounce<T extends (...args: any[]) => any>(
  fn: T,
  wait?: number,
  options?: DebounceSettings,
): T & { cancel(): void; flush(): ReturnType<T> | undefined };

// Implementation
export function debounce(fn: any, wait = DEFAULT_DEBOUNCE_DELAY, options: any = {}) {
  if ('onlyResolvesLast' in options && options.onlyResolvesLast === true) {
    return awesomeDebouncePromise(fn, wait, options);
  }
  return lodashDebounce(fn, wait, options);
}

/**
 * Legacy debounceAsync function - wraps awesome-debounce-promise for backwards compatibility
 *
 * @param asyncFunction - The async function to debounce
 * @param debounceTime - Delay in milliseconds (default: 250ms)
 * @param options - Awesome debounce options
 * @returns Debounced version of the function
 *
 * @deprecated Use debounce() with { onlyResolvesLast: true } when you need race-safe debouncing
 */
export function debounceAsync<T extends (...args: any[]) => Promise<any>>(
  asyncFunction: T,
  debounceTime = DEFAULT_DEBOUNCE_DELAY,
  options: Partial<AwesomeDebounceOptions> = { onlyResolvesLast: false },
): T & { cancel(): void } {
  return awesomeDebouncePromise(asyncFunction, debounceTime, options) as T & { cancel(): void };
}
