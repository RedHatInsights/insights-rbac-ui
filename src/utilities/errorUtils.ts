/**
 * Extracts user-friendly error messages from API error responses
 * Filters out technical messages like axios errors and provides fallbacks
 */
export const extractErrorMessage = (error: any): string => {
  // Default fallback message
  const defaultMessage = 'Something went wrong. Please try again.';

  if (!error) {
    return defaultMessage;
  }

  if (error.errors && error.errors.length > 0) {
    // API error format: { errors: [{ detail: "message" }] }
    return error.errors[0].detail || defaultMessage;
  }

  if (error.detail) {
    // Direct error format: { detail: "message" }
    return error.detail;
  }

  if (error.message && !error.message.includes('status code')) {
    // Only use error.message if it's not an axios error
    return error.message;
  }

  // Fallback for any other case (including axios errors)
  return defaultMessage;
};
