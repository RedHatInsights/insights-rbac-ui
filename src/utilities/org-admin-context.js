import { createContext } from 'react';

const OrgAdminContext = createContext(false);

export default OrgAdminContext;
