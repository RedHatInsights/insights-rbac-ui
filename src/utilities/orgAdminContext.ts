import { createContext } from 'react';

const OrgAdminContext = createContext<boolean>(false);

export default OrgAdminContext;
