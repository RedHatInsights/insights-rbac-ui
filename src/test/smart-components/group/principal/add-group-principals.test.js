import React from 'react';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { shallowToJson } from 'enzyme-to-json';

import { MemoryRouter } from 'react-router-dom';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';
import { RBAC_API_BASE } from '../../../../utilities/constants';
import AddGroupMembers from '../../../../smart-components/group/principal/add-group-members';

describe('<AddGroupMembers />', () => {
  let initialState;
  let responseBody;
  const middlewares = [thunk, promiseMiddleware, notificationsMiddleware()];
  let mockStore;

  beforeEach(() => {
    initialState = {
      groupReducer: {
        groups: {
          data: [
            {
              uuid: '123',
              name: 'SampleGroup',
            },
          ],
        },
      },
      userReducer: {
        selectedUser: {},
        isUserDataLoading: false,
        users: {
          data: [],
          meta: {
            count: 0,
          },
        },
      },
    };
    responseBody = {
      data: [
        {
          username: 'test_name',
          email: 'testmail@redhat.com',
          first_name: 'test',
          last_name: 'test',
          is_active: true,
        },
      ],
      meta: {
        count: 1,
        limit: 10,
        offset: undefined,
      },
    };
    mockStore = configureStore(middlewares);
  });

  it('should render correctly', () => {
    const store = mockStore(initialState);
    const wrapper = shallow(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/groups/detail/test-group/members/add_members']}>
          <AddGroupMembers></AddGroupMembers>
        </MemoryRouter>
      </Provider>
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should fetchUsers on load correctly', (done) => {
    const store = mockStore(initialState);
    apiClientMock.get(`${RBAC_API_BASE}/principals/?limit=10&sort_order=asc`, mockOnce({ body: responseBody }));
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/groups/detail/test-group/members/add_members']}>
          <AddGroupMembers></AddGroupMembers>
        </MemoryRouter>
      </Provider>
    );
    const expectedPayload = [
      expect.objectContaining({ type: 'FETCH_USERS_PENDING' }),
      expect.objectContaining({
        type: 'FETCH_USERS_FULFILLED',
        payload: responseBody,
      }),
    ];
    setImmediate(() => {
      expect(store.getActions()).toEqual(expectedPayload);
      expect(wrapper.find(AddGroupMembers)).toHaveLength(1);
      done();
    });
  });

  it('should show a notification on cancel', (done) => {
    const store = mockStore(initialState);
    apiClientMock.get(`${RBAC_API_BASE}/principals/?limit=10&sort_order=asc`, mockOnce({ body: responseBody }));
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/groups/detail/test-group/members/add_members']}>
          <AddGroupMembers></AddGroupMembers>
        </MemoryRouter>
      </Provider>
    );
    const expectedPayload = [
      expect.objectContaining({ type: 'FETCH_USERS_PENDING' }),
      expect.objectContaining({ type: '@@INSIGHTS-CORE/NOTIFICATIONS/ADD_NOTIFICATION' }),
      expect.objectContaining({
        type: 'FETCH_USERS_FULFILLED',
        payload: responseBody,
      }),
    ];
    wrapper.find('.pf-m-link').simulate('click');
    wrapper.update();
    setImmediate(() => {
      expect(store.getActions()).toEqual(expectedPayload);
      done();
    });
  });
});
