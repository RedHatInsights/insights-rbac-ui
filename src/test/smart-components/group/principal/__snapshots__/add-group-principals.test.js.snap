// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddGroupMembers /> should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        "/groups/detail/test-group/members/add_members",
      ]
    }
  >
    <withRouter(Connect(AddGroupMembers)) />
  </MemoryRouter>
</ContextProvider>
`;
